src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -246,96 +253,111 @@ object RuleUtils {
    * @param originalPlan Original plan.
-   * @param planWithIndex Replaced plan with index.
    * @param filesAppended Appended files to the source relation.
-   * @return Consolidated plan of planWithIndex and a complement plan for filesAppended
+   * @return Linear logical plan for appended files.
    */
-  private def getComplementIndexPlan(
+  private def transformPlanWithAppendedFiles(
       spark: SparkSession,
-      index: IndexLogEntry,
+      indexSchema: StructType,
       originalPlan: LogicalPlan,
-      planWithIndex: LogicalPlan,
       filesAppended: Seq[Path]): LogicalPlan = {
-    // 1) Replace the location of LogicalRelation with appended files
-    val complementIndexPlan = originalPlan transformDown {
-      case baseRelation @ LogicalRelation(
-            fsRelation @ HadoopFsRelation(location: PartitioningAwareFileIndex, _, _, _, _, _),
-            baseOutput,
-            _,
-            _) =>
-        // Set the same output schema with the index plan to merge them using BucketUnion
+    // Replace the location of LogicalRelation with appended files.
+    val planForAppended = originalPlan transformDown {
+      case baseRelation @ LogicalRelation(fsRelation: HadoopFsRelation, baseOutput, _, _) =>
+        // Set the same output schema with the index plan to merge them using BucketUnion.
         val updatedOutput =
-          baseOutput.filter(attr => index.schema.fieldNames.contains(attr.name))
-
-        if (filesAppended.nonEmpty) {
-          val newLocation =
-            new InMemoryFileIndex(spark, filesAppended, Map(), None)
-          val newRelation =
-            fsRelation.copy(
-              location = newLocation,
-              dataSchema = StructType(index.schema.filter(baseRelation.schema.contains(_))))(
-              spark)
-          baseRelation.copy(relation = newRelation, output = updatedOutput)
-        } else {
-          baseRelation
-        }
+          baseOutput.filter(attr => indexSchema.fieldNames.contains(attr.name))
+        val newLocation =
+          new InMemoryFileIndex(spark, filesAppended, Map(), None)
+        val newRelation =
+          fsRelation.copy(
+            location = newLocation,
+            dataSchema = StructType(indexSchema.filter(baseRelation.schema.contains(_))))(spark)
+        baseRelation.copy(relation = newRelation, output = updatedOutput)
     }
-    assert(!originalPlan.equals(complementIndexPlan))
+    assert(!originalPlan.equals(planForAppended))
+    planForAppended
+  }
 
+  /**
+   * Transform the plan to perform on-the-fly Shuffle the data based on bucketSpec.
+   *
+   * Pre-requisites
+   * - The plan should be linear and include 1 LogicalRelation.
+   *
+   * @param bucketSpec Bucket specification used for Shuffle.
+   * @param plan Plan to be shuffled.
+   * @return Transformed plan by injecting on-the-fly shuffle with given bucket specification.
+   */
+  private[rules] def shuffleUsingIndexSpec(
```

Comment:
How about:

```suggestion
  private[rules] def transformPlanToShuffleUsingIndexSpec(
```

I'll leave it to you. Either looks good to me. This comment is only to stay consistent.